test_runner/batch_others/test_readonly_node.py

```diff
@@ -84,8 +96,19 @@ def test_readonly_node(zenith_simple_env: ZenithEnv):
     assert cur.fetchone() == (1, )
 
     # Create node at pre-initdb lsn
-    try:
+    with pytest.raises(Exception, match='extracting base backup failed'):
+        # compute node startup with invalid LSN should fail
         env.zenith_cli(["pg", "start", "test_branch_preinitdb", "test_readonly_node@0/42"])
-        assert False, "compute node startup with invalid LSN should have failed"
-    except Exception:
-        print("Node creation with pre-initdb LSN failed (as expected)")
+
+    # check that we cannot create branch based on garbage collected data
```

Comment:
copy-pasted comment: we're not creating a branch here, but an "anchored" read-only node. Same below

Reply:
Hm, but what is the difference between not advanced branch and readonly node? It also calls branch_timeline function internally

Reply:
Removed the duplicate though

Reply:
>  It also calls branch_timeline function internally

No, it doesn't. Or am I missing something?

Reply:
Error originates from branch_timeline so there wouldn't be an error if it is not called :) Also nothing is found by grepping rust source for `readonly`, so I assume that readonly node is just a branch without changes since the branchpoint

Reply:
No, a readonly node doesn't create a branch. You can have many read-only nodes on the same branch, at different LSNs, and there is no trace of those nodes in the page server, except for the GetPage requests that they make.

Reply:
> Error originates from branch_timeline so there wouldn't be an error if it is not called :) 

Ok, this doesn't count as a complete fix for https://github.com/zenithdb/zenith/issues/95 then. We should have a similar check somewhere in get_page_at_lsn() or wait_lsn(), too.

Reply:
> and there is no trace of those nodes in the page server, except for the GetPage requests that they make.

IMO this is very fragile then. I do not think that user expects that this read only node starts throwing errors after some time. IMO when this happens it is a bug in our code. I can think about it as something like debug assert only

Reply:
The comment here still needs updating

Reply:
This checks branching case, so IMO the comment is valid

Reply:
And error during get_page_at_lsn is checked in a unittest

Reply:
> This checks branching case, so IMO the comment is valid

Oh, ok. Perhaps this should be in the `branch_behind` test rather than here, then?

Looking at 'test_branch_behind.py', does this PR fix the FIXME near the end of that file?

Reply:
> Perhaps this should be in the branch_behind test rather than here

Good idea, this is indeed a better place for it

> does this PR fix the FIXME near the end of that file?

Sort of. If there is a GC run before the check it will fail with `Branch start LSN 0/48 is earlier than latest GC horizon 0/3EC8088 (we might've already garbage collected needed data)`. So if there is no GC run it will still work. What do you think about hardcoding this lsn (or remembering that lsn after initdb if it can change) and adding a new check to avoid this?

Also note that as we discussed there is an escape hatch. If you run `zenith pg start foo@<some gcd lsn>` node will try to start but it immediately fails because of the check in get_page_at_lsn added in this PR

Reply:
> Good idea, this is indeed a better place for it

Done